tests/bmat.test.cc

```diff
@@ -0,0 +1,151 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+
+#include "catch.hpp"
+
+#include "../src/bmat.h"
+
+#define BMAT_REPORT false
+
+using namespace libsemigroups;
+
+TEST_CASE("BMat 01: transpose", "[quick][bmat][01]") {
+  BMat8 bm1(0);
+  REQUIRE(bm1.transpose() == bm1);
+
+  BMat8 bm2({{1, 1}, {0, 1}});
+  REQUIRE(bm2.transpose() == BMat8({{1, 0}, {1, 1}}));
+
+  BMat8 bm3({{0, 0, 0, 1, 0, 0, 1, 1},
+             {1, 1, 1, 1, 1, 1, 0, 1},
+             {0, 1, 1, 1, 0, 1, 0, 1},
+             {1, 1, 0, 1, 1, 1, 1, 1},
+             {0, 0, 1, 0, 0, 1, 1, 1},
+             {1, 1, 0, 0, 0, 0, 0, 1},
+             {0, 1, 0, 0, 0, 0, 1, 1},
+             {0, 1, 1, 1, 1, 0, 1, 0}});
+
+  REQUIRE(bm3.transpose() == BMat8({{0, 1, 0, 1, 0, 1, 0, 0},
+                                    {0, 1, 1, 1, 0, 1, 1, 1},
+                                    {0, 1, 1, 0, 1, 0, 0, 1},
+                                    {1, 1, 1, 1, 0, 0, 0, 1},
+                                    {0, 1, 0, 1, 0, 0, 0, 1},
+                                    {0, 1, 1, 1, 1, 0, 0, 0},
+                                    {1, 0, 0, 1, 1, 0, 1, 1},
+                                    {1, 1, 1, 1, 1, 1, 1, 0}}));
+}
+
+TEST_CASE("BMat 02: multiplication", "[quick][bmat][02]") {
+  BMat8 bm({{0, 0, 0, 1, 0, 0, 1, 1},
+            {1, 1, 1, 1, 1, 1, 0, 1},
+            {0, 1, 1, 1, 0, 1, 0, 1},
+            {1, 1, 0, 1, 1, 1, 1, 1},
+            {0, 0, 1, 0, 0, 1, 1, 1},
+            {1, 1, 0, 0, 0, 0, 0, 1},
+            {0, 1, 0, 0, 0, 0, 1, 1},
+            {0, 1, 1, 1, 1, 0, 1, 0}});
+
+  BMat8 tmp = bm * bm.one();
+  REQUIRE(tmp == bm);
+  REQUIRE(tmp == bm * bm.one());
+
+  tmp = bm.one() * bm;
+  REQUIRE(tmp == bm);
+  REQUIRE(tmp == bm.one() * bm);
+
+  tmp = bm * BMat8(0);
+  REQUIRE(tmp == BMat8(0));
+
+  BMat8 bm2({{0, 0, 0, 1, 0, 0, 1, 1},
+             {0, 0, 1, 0, 0, 1, 0, 1},
+             {1, 1, 0, 0, 1, 1, 0, 1},
+             {1, 1, 0, 0, 0, 0, 0, 1},
+             {0, 1, 0, 0, 0, 0, 1, 1},
+             {0, 1, 0, 1, 1, 1, 1, 1},
+             {0, 1, 0, 1, 0, 1, 0, 1},
+             {0, 1, 0, 0, 0, 0, 1, 0}});
+
+  tmp = bm * bm2;
+
+  BMat8 bm3({{1, 1, 0, 1, 0, 1, 1, 1},
+             {1, 1, 1, 1, 1, 1, 1, 1},
+             {1, 1, 1, 1, 1, 1, 1, 1},
+             {1, 1, 1, 1, 1, 1, 1, 1},
+             {1, 1, 0, 1, 1, 1, 1, 1},
+             {0, 1, 1, 1, 0, 1, 1, 1},
+             {0, 1, 1, 1, 0, 1, 1, 1},
+             {1, 1, 1, 1, 1, 1, 1, 1}});
+
+  REQUIRE(tmp == bm3);
+  REQUIRE(tmp == bm * bm2);
```

Comment:
Should probably have more tests for multiplication, but I can't see a decent / non-trivial way of testing a lot of multiplications directly. The row space basis test did include a large number of multiplications.

Reply:
Probably but I am in the process of adding tests for in the `template` branch that will perform millions of multiplications, so maybe don't worry about this for now.